### PR TITLE
CRM-383 Add in job and code to update contributions CiviCampaign (Cam…

### PR DESCRIPTION
**Pull request by seamuslee001**
…paignGroup) Value when the linked contribution page(Campaign) link is updated

@monishdeb @joemurray

This aims to get around some of those intepretation issues that have been cropping up with CH around when campaign groups are linked to campaigns and the affects they have on the report

At a high level this adds in a new job (need to be added to scheduled jobs) that will update the linked contributions to the new campaign group value which then should fix the reports

More detailed is that this adds api wrapper on the ContributionPage.create API and post process on the contributionPage settings form which triggers the process of updating the campaign value by adding them into a batch table a ala with the chfinds

This PR also changes the class name of the utils file and fixes the indendenting in it so it doesn't clash with the auto generated class in the civix file

### CODE REVIEW

**Comment by monishdeb** [comment]
@seamuslee001 @JoeMurray after reviewing the patch let me present my opinion how we can avoid having a separate table and job, to store the queue for contributions which need to be later processed with new campaign groups whenever mapping between campaign and campaign group changes. This approach is similar to what we are already doing for outdated contributions on CHFund and Fund mapping change. Here's my thought: 
1. Instead of having a separate table ```civicrm_ch_campaign_contribution_batch``` how about we modify the existing table [civicrm_ch_contribution_batch](https://github.com/CanadaHelps/chfunds/blob/master/sql/auto_install.sql#L98) to add an new column say ```campaign_id``` ? 
2. The above step will help us to rely on the existing job [UpdateCHContribtuions](https://github.com/CanadaHelps/chfunds/blob/master/api/v3/Job/UpdateCHContributions.php#L28). Only thing we need to do to is to bring some changes in the code so the Job will be able to process contributions both for outdated campaign group and funds. So the Job function will more look like this: 
```php
function civicrm_api3_job_update_c_h_contributions($params) {
  $batchSize = CRM_Utils_Array::value('batch_size', $params, 1000);
  $dao = CRM_Core_DAO::executeQuery("SELECT * FROM civicrm_ch_contribution_batch GROUP BY contribtution_id LIMIT 0, $batchSize ");
// NOTE: due to group by instead of having two rows, one for campaign group and other for fund, we can still get 1 row
  while($dao->fetch()) {
     $params = [
        'id' => $dao->contribution_id,
      ];
      if (!empty($dao->campaign_id)) {
        $params['campaign_id'] = $dao->campaign_id;
      }
     elseif($dao->campaign_id === 0) { 
       //edge case when campaign group is suppose to be empty as per new mapping
        $params['campaign_id'] = '';
      }
      if (!empty($dao->financial_type_id)) {
         $params['financial_type_id'] = $dao->financial_type_id;
      }
      civicrm_api3('Contribution', 'create', $params);
    CRM_Core_DAO::executeQuery("DELETE FROM civicrm_ch_contribution_batch WHERE contribution_id = " . $dao->contribtuion_id);
  }
  return civicrm_api3_create_success();
}
```
Now the advantages of having a single job to process outdated contributions would be : 
a) Its always expensive to have two cron jobs runing in the background instead we can modify existing job to update contribtuion for both campaign group and fund. 
b) Merge two processes into one, suppose theres a situation where both the batch table hold the entry for same contribtuion and we are calling the Contribution.update API twice which can be avoided by single API call as shown above.

3. Lets have a separate fn as in here [updateContributionCampaign](https://github.com/CanadaHelps/chfunds/pull/21/files#diff-7367bcd3cbb5bf463c590a52d70e6d08R116) but we need to modify it a bit to use the existing batch table to populate rows with contribtuion entries which need to be processed later. 

4. And of course modify the UT as per the above approach. 

What do you think guys?

**Comment by seamuslee001** [comment]
@monishdeb I have adapated that now, If you think this is ok then we should deploy it to test sites to QA

File: CRM/Chfunds/Upgrader.php [hunk 1]
```diff
@@ -38,6 +38,19 @@ public function upgrade_1300() {
     return TRUE;
   }
 
+  public function upgrade_1400() {
+    $this->ctx->log->info('Applying update 1.4');
+    $sql = "
+      ALTER TABLE `civicrm_ch_contribution_batch` ADD COLUMN `campaign_id` INT NOT NULL
```
Comment (monishdeb) [inline CRM/Chfunds/Upgrader.php]: @seamuslee001 can you also add alter query to set DEFAULT NULL property for ch_fund and fund column ? i.e.

```
"ALTER TABLE `civicrm_ch_contribution_batch` ADD COLUMN `campaign_id` INT NOT NULL,
 CHANGE `ch_fund` `ch_fund` VARCHAR(255) CHARACTER SET utf8 COLLATE utf8_unicode_ci NULL DEFAULT NULL,
 CHANGE `fund` `fund` INT(11) NULL DEFAULT NULL;"
```

**Comment by monishdeb** [comment]
I encountered the following issues: 

1. On submitting contribution page setting page on selecting Campaign Group results into : 
Error: Using $this when not in object context in chfunds_civicrm_postProcess() (line 350 of /Users/monish/src/civi-extensions/chfunds/chfunds.php).

2. On visiting the Financial Type page it ends up into a fatal error:
Error: Call to undefined method CRM_Chfunds_ExtensionUtil::getCHFundsByFinancialType() in chfunds_civicrm_pageRun() (line 206 of /Users/monish/src/civi-extensions/chfunds/chfunds.php).

3. campaign_id column is not set Default to null on update column.

Fixed all the three issues in this [commit](https://github.com/CanadaHelps/chfunds/pull/21/commits/2012c32c1dd372f068ec1432ec26166de1fa7e72) and updated the PR.

**Comment by monishdeb** [comment]
Test Results
---------------- 

Extension Upgrade : 
After upgrading ext to 1.4, the following changes are made into civicrm_ch_contribution_batch table: 
1. New column campaign_id with Default as NULL is added 
2. ch_fund and fund is set with Default NULL 
3. Unique key is added on contribution key. 
All these schema changes are correct and needed. 

Test CASE 1: 
1. Created two Campaign groups - Campaign test1 and Campaign test2
2. Saved the setting of existing Campaign (say A) with campaign group - Campaign test1
3. Used online Campaign page A to record live contributions. 
4. Checked the contribtuion view and confirmed that it is associated with 'Campaign Test1' Campaign group
5. Went back to campaign setting page and changed the Campaign Group to 'Campaign test2' and saved.
6. Changed the CH Fund mapping with the Fund used in the contribution. 

RESULT:
As a result of change in mapping between Campaign and Campaign Group & between Fund and CH Fund, I found an entry in 
 civicrm_ch_contribution_batch with contribution ID and new campaign group ID of 'Campaign test2' and new CHFund value against Fund. Executing the job, successfully updated the contribution to link with new campaign group and CHFund - correct

TEST CASE 2:

1. Continued the test from step 5 and on campaign setting page, changed the  Campaign Group selection to none and saved

RESULT 
As a result of the change in mapping between Campaign and Campaign Group, I found an entry in  civicrm_ch_contribution_batch with contribution ID and campaign ID is set to 0. Executing the job, successfully updated the contribution and removed the associated campaign group value.  - correct


Tested the unit test and it captures the scenario well: 
```
$ phpunit5 --filter=testUpdateContributionCampaignGroupFromCampaign tests/phpunit/api/v3/ChContribution/CreateTest.php 
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

.                                                                   1 / 1 (100%)

Time: 13.72 seconds, Memory: 24.57MB
OK (1 test, 27 assertions)
```

**Pull request merged**